write-fonts/src/tables/variations.rs

```diff
@@ -366,6 +369,538 @@ impl Tuple {
     }
 }
 
+#[derive(Copy, Clone, Debug)]
+pub enum IupError {
+    DeltaCoordLengthMismatch {
+        num_deltas: usize,
+        num_coords: usize,
+    },
+    NotEnoughCoords(usize),
+    CoordEndsMismatch {
+        num_coords: usize,
+        expected_num_coords: usize,
+    },
+    AchievedInvalidState,
+}
+
+trait WrappingGet<T> {
+    fn wrapping_next(&self, idx: usize) -> &T;
+    fn wrapping_prev(&self, idx: usize) -> &T;
+}
+
+impl<T> WrappingGet<T> for &[T] {
+    fn wrapping_next(&self, idx: usize) -> &T {
+        &self[match idx {
+            _ if idx == self.len() - 1 => 0,
+            _ => idx + 1,
+        }]
+    }
+
+    fn wrapping_prev(&self, idx: usize) -> &T {
+        &self[match idx {
+            _ if idx == 0 => self.len() - 1,
+            _ => idx - 1,
+        }]
+    }
+}
+
+/// Check if IUP _might_ be possible. If not then we *must* encode the value at this index.
+///
+/// <https://github.com/fonttools/fonttools/blob/6a13bdc2e668334b04466b288d31179df1cff7be/Lib/fontTools/varLib/iup.py#L238-L290>
+fn must_encode_at(deltas: &[Vec2], coords: &[Point], tolerance: f64, at: usize) -> bool {
+    let ld = *deltas.wrapping_prev(at);
+    let d = deltas[at];
+    let nd = *deltas.wrapping_next(at);
+    let lc = *coords.wrapping_prev(at);
+    let c = coords[at];
+    let nc = *coords.wrapping_next(at);
+
+    for axis in [Axis2D::X, Axis2D::Y] {
+        let (ldj, lcj) = (ld.get(axis), lc.get(axis));
+        let (dj, cj) = (d.get(axis), c.get(axis));
+        let (ndj, ncj) = (nd.get(axis), nc.get(axis));
+        let (c1, c2, d1, d2) = if lcj <= ncj {
+            (lcj, ncj, ldj, ndj)
+        } else {
+            (ncj, lcj, ndj, ldj)
+        };
+        // If the two coordinates are the same, then the interpolation
+        // algorithm produces the same delta if both deltas are equal,
+        // and zero if they differ.
+        match (c1, c2) {
```

Comment:
Nothing, I just thought it read better than if/else if/etc.

Reply:
Also hopefully makes it clearer which paths are exclusive. My first translation actually had an error where you could take two of the paths because I misread the python slightly.